File: MultiAgent/movement.py
l = 1
n = 30
pi = 0.05
v = l / 20
from random import random, uniform
import pandas as pd 
agentes =  pd.DataFrame()
agentes['x'] = [uniform(0, l) for i in range(n)]
agentes['y'] = [uniform(0, l) for i in range(n)]
agentes['dx'] = [uniform(-v, v) for i in range(n)]
agentes['dy'] = [uniform(-v, v) for i in range(n)]
agentes['estado'] = ['S' if random() > pi else 'I' for i in range(n)]
 
import matplotlib.pyplot as plt
from math import floor, log, sqrt
 
epidemia = []
r = 0.1
tmax = 30
digitos = floor(log(tmax, 10)) + 1
for tiempo in range(tmax):
    conteos = agentes.estado.value_counts()
    infectados = conteos.get('I', 0)
    epidemia.append(infectados)
    if infectados == 0:
        break
    contagios = [False for i in range(n)]
    for i in range(n): # contagios
        a1 = agentes.iloc[i]
        if a1.estado == 'I':
            for j in range(n):
                a2 = agentes.iloc[j]
                if a2.estado == 'S':
                    dx = a1.x - a2.x
                    dy = a1.y - a2.y
                    d = sqrt(dx**2 + dy**2)
                    if d < r:
                        p = (r - d) / r
                        if random() < p:
                            contagios[j] = True
    for i in range(n): # movimientos
        a = agentes.iloc[i]
        if contagios[i]:
            agentes.at[i, 'estado'] = 'I'
        x = a.x + a.dx
        y = a.y + a.dy
        x = x if x < l else x - l
        y = y if y < l else y - l
        x = x if x > 0 else x + l
        y = y if y > 0 else y + l
        agentes.at[i, 'x'] = x
        agentes.at[i, 'y'] = y
    fig = plt.figure()
    ax = plt.subplot(1, 1, 1)
    plt.xlim(0, l)
    plt.ylim(0, l)
    c = {'I': 'r', 'S': 'g', 'R': 'orange'}
    m = {'I': 'o', 'S': 's', 'R': '2'}
    for e, d in agentes.groupby('estado'):
        if len(d) > 0:
            ax.scatter(d.x, d.y, c = c[e], marker = m[e])
    plt.xlabel('x')
    plt.xlabel('y')
    plt.title('Paso {:d}'.format(tiempo + 1))
    fig.savefig('p6p_t' + format(tiempo, '0{:d}'.format(digitos)) + '.png')
    plt.close()
    print(epidemia)

File: MultiAgent/movement.R
l <- 1
n <- 50
pi <- 0.05
v <- l / 20
agentes <- data.frame(x = double(), y = double(),
                      dx = double(), dy = double(),
                      estado  = character())
for (i in 1:n) {
    e <- "S"
    if (runif(1) < pi) {
        e <- "I"
    }
    agentes <- rbind(agentes, data.frame(x = runif(1, 0, l),
                                         y = runif(1, 0, l),
                                         dx = runif(1, -v, v),
                                         dy = runif(1, -v, v),
                                         estado = e))
}
levels(agentes$estado) <- c(levels(agentes$estado), "R")
 
epidemia <- integer()
r <- 0.1
tmax <- 30
digitos <- floor(log(tmax, 10)) + 1
for (tiempo in 1:tmax) {
    infectados <- dim(agentes[agentes$estado == "I",])[1]
    epidemia <- c(epidemia, infectados)
    if (infectados == 0) {
        break
    }
    contagios <- rep(FALSE, n)
    for (i in 1:n) { # posibles contagios
        a1 <- agentes[i, ]
        if (a1$estado == "I") { # desde los infectados
            for (j in 1:n) {
                if (!contagios[j]) { # aun sin contagio
                    a2 <- agentes[j, ]
                    if (a2$estado == "S") { # hacia los susceptibles
                        dx <- a1$x - a2$x
                        dy <- a1$y - a2$y
                        d <- sqrt(dx^2 + dy^2)
                        if (d < r) { # umbral
                            p <- (r - d) / r
                            if (runif(1) < p) {
                                contagios[j] <- TRUE
                            }
                        }
                    }
                }
            }
        }
    }
    for (i in 1:n) { # movimientos
        a <- agentes[i, ]
        if (contagios[i]) {
            a$estado <- "I"
        }
        a$x <- a$x + a$dx
        a$y <- a$y + a$dy
        if (a$x > l) {
            a$x <- a$x - l
        }
        if (a$y > l) {
            a$y <- a$y - l
        }
        if (a$x < 0) {
            a$x <- a$x + l
        }
        if (a$y < 0) {
            a$y <- a$y + l
        }
        agentes[i, ] <- a
    }
    aS <- agentes[agentes$estado == "S",]
    aI <- agentes[agentes$estado == "I",]
    aR <- agentes[agentes$estado == "R",]
    tl <- paste(tiempo, "", sep="")
    while (nchar(tl) < digitos) {
        tl <- paste("0", tl, sep="")
    }
    salida <- paste("p6_t", tl, ".png", sep="")
    tiempo <- paste("Paso", tiempo)
    png(salida)
    plot(l, type="n", main=tiempo, xlim=c(0, l), ylim=c(0, l), xlab="x", ylab="y")
    if (dim(aS)[1] > 0) {
        points(aS$x, aS$y, pch=15, col="chartreuse3", bg="chartreuse3")
    }
    if (dim(aI)[1] > 0) {
        points(aI$x, aI$y, pch=16, col="firebrick2", bg="firebrick2")
    }
    if (dim(aR)[1] > 0) {
        points(aR$x, aR$y, pch=17, col="goldenrod", bg="goldenrod")
    }
    graphics.off()
}
cat(epidemia, "\n")
